src/amqp-client/channel.cr

```diff
@@ -256,24 +280,35 @@ class AMQP::Client
       if @confirm_mode
         @confirm_lock.synchronize do
           msgid = @confirm_id &+= 1_u64
-          @unconfirmed_publishes.push msgid
+          @unconfirmed_publishes.push({msgid, blk})
           msgid
         end
       else
+        raise ArgumentError.new("On confirm block received but channel is not in confirm mode") if blk
         0_u64
       end
     end
 
     def basic_publish_confirm(msg, exchange, routing_key = "", mandatory = false, immediate = false, props properties = Properties.new) : Bool
       confirm_select
-      basic_publish(msg, exchange, routing_key, mandatory, immediate, properties)
-      wait_for_confirms
+      done = ::Channel(Nil).new
+      basic_publish(msg, exchange, routing_key, mandatory, immediate, properties) do
+        done.send nil
+      end
+      done.receive
```

Comment:
This is probably very inefficient, creating a new `Channel` (on the heap) for each publish..